File: backend/src/models/question.py
from sqlalchemy import Column, Integer, String, Text, Boolean, Enum, ForeignKey, JSON
from sqlalchemy.orm import relationship
import enum
from . import Base, TimestampMixin, SoftDeleteMixin


class QuestionType(str, enum.Enum):
    """Question type enumeration."""
    MULTIPLE_CHOICE = "multiple_choice"
    FREE_TEXT = "free_text"
    DATABASE_DROPDOWN = "database_dropdown"
    PERSON = "person"
    DATE = "date"


class QuestionGroup(Base, TimestampMixin, SoftDeleteMixin):
    """Question group model."""
    
    __tablename__ = "question_groups"
    
    id = Column(Integer, primary_key=True, index=True)
    name = Column(String(255), nullable=False)
    description = Column(Text, nullable=True)
    identifier = Column(String(100), unique=True, nullable=False, index=True)
    display_order = Column(Integer, default=0, nullable=False)
    
    # Question logic stored as JSON (list of question items with conditionals)
    # Structure: [{ type: 'question', questionId: 123 }, { type: 'conditional', ifIdentifier: 'prev_q', value: 'yes', nestedItems: [...] }]
    question_logic = Column(JSON, nullable=True)
    
    # Relationships
    questions = relationship(
        "Question",
        back_populates="question_group",
        cascade="all, delete-orphan",
        order_by="Question.display_order"
    )
    
    def __repr__(self) -> str:
        return f"<QuestionGroup(id={self.id}, name='{self.name}', identifier='{self.identifier}')>"
    
    def to_dict(self) -> dict:
        """Convert question group to dictionary."""
        return {
            "id": self.id,
            "name": self.name,
            "description": self.description,
            "identifier": self.identifier,
            "display_order": self.display_order,
            "question_logic": self.question_logic,
            "created_at": self.created_at.isoformat(),
            "updated_at": self.updated_at.isoformat(),
            "is_active": self.is_active,
            "question_count": len(self.questions) if self.questions else 0,
        }


class Question(Base, TimestampMixin, SoftDeleteMixin):
    """Question model."""
    
    __tablename__ = "questions"
    
    id = Column(Integer, primary_key=True, index=True)
    question_group_id = Column(Integer, ForeignKey("question_groups.id"), nullable=False, index=True)
    question_text = Column(Text, nullable=False)
    question_type = Column(String(50), nullable=False)  # Validated by Pydantic schema
    identifier = Column(String(100), nullable=False, index=True)
    repeatable = Column(Boolean, default=False, nullable=False)
    repeatable_group_id = Column(String(100), nullable=True)  # ID to group repeatable questions together
    display_order = Column(Integer, default=0, nullable=False)
    is_required = Column(Boolean, default=True, nullable=False)
    help_text = Column(Text, nullable=True)
    
    # For multiple choice questions
    options = Column(JSON, nullable=True)  # List of options: [{"value": "male", "label": "Male"}, ...]
    
    # For database dropdown questions
    database_table = Column(String(100), nullable=True)
    database_value_column = Column(String(100), nullable=True)
    database_label_column = Column(String(100), nullable=True)
    
    # For person type questions
    person_display_mode = Column(String(20), nullable=True)  # 'autocomplete' or 'dropdown'
    
    # For date type questions
    include_time = Column(Boolean, default=False, nullable=True)  # Whether to include time of day
    
    # Validation rules
    validation_rules = Column(JSON, nullable=True)  # {"min_length": 5, "max_length": 100, "pattern": "regex"}
    
    # Relationships
    question_group = relationship("QuestionGroup", back_populates="questions")
    
    def __repr__(self) -> str:
        qt = self.question_type.value if hasattr(self.question_type, 'value') else self.question_type
        return f"<Question(id={self.id}, identifier='{self.identifier}', type='{qt}')>"
    
    def to_dict(self) -> dict:
        """Convert question to dictionary."""
        qt = self.question_type.value if hasattr(self.question_type, 'value') else self.question_type
        return {
            "id": self.id,
            "question_group_id": self.question_group_id,
            "question_text": self.question_text,
            "question_type": qt,
            "identifier": self.identifier,
            "repeatable": self.repeatable,
            "repeatable_group_id": self.repeatable_group_id,
            "display_order": self.display_order,
            "is_required": self.is_required,
            "help_text": self.help_text,
            "options": self.options,
            "database_table": self.database_table,
            "database_value_column": self.database_value_column,
            "database_label_column": self.database_label_column,
            "person_display_mode": self.person_display_mode,
            "include_time": self.include_time,
            "validation_rules": self.validation_rules,
            "created_at": self.created_at.isoformat(),
            "updated_at": self.updated_at.isoformat(),
            "is_active": self.is_active,
        }

File: go-backend/internal/models/question.go
package models

import (
	"encoding/json"
	"time"
)

type QuestionGroup struct {
	ID             int              `gorm:"primaryKey" json:"id"`
	Name           string           `gorm:"size:255;not null" json:"name"`
	Description    *string          `gorm:"type:text" json:"description"`
	Identifier     string           `gorm:"size:100;uniqueIndex;not null" json:"identifier"`
	DisplayOrder   int              `gorm:"default:0;not null" json:"display_order"`
	QuestionLogic  *json.RawMessage `gorm:"type:jsonb" json:"question_logic"`
	CollapsedItems *json.RawMessage `gorm:"type:jsonb" json:"collapsed_items"`
	IsActive       bool             `gorm:"default:true;not null" json:"is_active"`
	CreatedAt      time.Time        `gorm:"not null" json:"created_at"`
	UpdatedAt      time.Time        `gorm:"not null" json:"updated_at"`

	Questions []Question `gorm:"foreignKey:QuestionGroupID;constraint:OnDelete:CASCADE" json:"questions,omitempty"`
}

func (QuestionGroup) TableName() string { return "question_groups" }

type QuestionGroupResponse struct {
	ID             int              `json:"id"`
	Name           string           `json:"name"`
	Description    *string          `json:"description"`
	Identifier     string           `json:"identifier"`
	DisplayOrder   int              `json:"display_order"`
	QuestionLogic  *json.RawMessage `json:"question_logic,omitempty"`
	CollapsedItems *json.RawMessage `json:"collapsed_items,omitempty"`
	IsActive       bool             `json:"is_active"`
	CreatedAt      time.Time        `json:"created_at"`
	UpdatedAt      time.Time        `json:"updated_at"`
	QuestionCount  int              `json:"question_count"`
	Questions      interface{}      `json:"questions,omitempty"`
}

type Question struct {
	ID                int              `gorm:"primaryKey" json:"id"`
	QuestionGroupID   int              `gorm:"index;not null" json:"question_group_id"`
	QuestionText      string           `gorm:"type:text;not null" json:"question_text"`
	QuestionType      string           `gorm:"size:50;not null" json:"question_type"`
	Identifier        string           `gorm:"size:100;index;not null" json:"identifier"`
	Repeatable        bool             `gorm:"default:false;not null" json:"repeatable"`
	RepeatableGroupID *string          `gorm:"size:100" json:"repeatable_group_id"`
	DisplayOrder      int              `gorm:"default:0;not null" json:"display_order"`
	IsRequired        bool             `gorm:"default:true;not null" json:"is_required"`
	HelpText          *string          `gorm:"type:text" json:"help_text"`
	Options           *json.RawMessage `gorm:"type:jsonb" json:"options"`
	DatabaseTable     *string          `gorm:"size:100" json:"database_table"`
	DatabaseValueCol  *string          `gorm:"column:database_value_column;size:100" json:"database_value_column"`
	DatabaseLabelCol  *string          `gorm:"column:database_label_column;size:100" json:"database_label_column"`
	PersonDisplayMode *string          `gorm:"size:20" json:"person_display_mode"`
	IncludeTime       *bool            `gorm:"default:false" json:"include_time"`
	ValidationRules   *json.RawMessage `gorm:"type:jsonb" json:"validation_rules"`
	IsActive          bool             `gorm:"default:true;not null" json:"is_active"`
	CreatedAt         time.Time        `gorm:"not null" json:"created_at"`
	UpdatedAt         time.Time        `gorm:"not null" json:"updated_at"`
}

func (Question) TableName() string { return "questions" }
